implementation/src/test/java/io/smallrye/mutiny/operators/MultiIfNoItemTest.java

```diff
@@ -0,0 +1,210 @@
+package io.smallrye.mutiny.operators;
+
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.TestException;
+import io.smallrye.mutiny.TimeoutException;
+import io.smallrye.mutiny.helpers.ParameterValidation;
+import io.smallrye.mutiny.helpers.test.AssertSubscriber;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.concurrent.Executors;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+
+public class MultiIfNoItemTest {
+
+    @Test
+    public void testResultWhenTimeoutIsNotReached() {
+        AssertSubscriber<Long> subscriber = AssertSubscriber.create(10);
+
+        Multi.createFrom().ticks().every(Duration.ofMillis(10))
+                .select().first(4)
+                .ifNoItem().after(Duration.ofMillis(20)).recoverWithCompletion()
+                .subscribe().withSubscriber(subscriber);
+
+        subscriber.awaitCompletion().assertItems(0L, 1L, 2L, 3L);
+    }
+
+    @Test
+    public void testTimeout() {
+        AssertSubscriber<Long> subscriber = AssertSubscriber.create();
+
+        Multi.createFrom().ticks().startingAfter(Duration.ofMillis(10)).every(Duration.ofMillis(20))
+                .ifNoItem().after(Duration.ofMillis(1)).fail()
+                .subscribe().withSubscriber(subscriber);
+
+        subscriber.awaitFailure();
+        assertThat(subscriber.getFailure()).isInstanceOf(TimeoutException.class);
+    }
+
+    @Test
+    public void testTimeoutOnFailure() {
+        AssertSubscriber<Long> subscriber = AssertSubscriber.create();
+
+        Multi.createFrom().<Long> emitter(e -> new Thread(() -> {
+            try {
+                Thread.sleep(10);
+            } catch (InterruptedException interruptedException) {
+                Thread.currentThread().interrupt();
+            }
+            e.fail(new TestException("boom"));
+        }).start())
+                .ifNoItem().after(Duration.ofMillis(1)).fail()
+                .subscribe().withSubscriber(subscriber);
+
+        subscriber.awaitFailure();
+        assertThat(subscriber.getFailure()).isInstanceOf(TimeoutException.class);
+    }
+
+    @Test
+    @Timeout(2)
+    public void testFailureBeforeTimeout() {
+        AssertSubscriber<Integer> subscriber = AssertSubscriber.create();
+
+        Multi.createFrom().<Integer> emitter(e -> new Thread(() -> {
+            try {
```

Comment:
You might simplify conditions like this one in tests using `Awaitability `, and speed up tests compared to having a hard thread sleep.

Reply:
Would adding a duration argument to the `subscriber.awaitFailure()` alleviate this issue? There is also a timeout annotation added to the test. I could be misunderstanding how `Awaitility` is supposed to be use.

Reply:
You'll find usages of the library in many places across the tests, as in:

```java
@RepeatedTest(10)
    public void testPredicateFailureWithWhilst() {
        AtomicInteger count = new AtomicInteger();
        AssertSubscriber<Integer> subscriber = Uni.createFrom().item(count::incrementAndGet)
                .repeat().whilst(v -> {
                    if (v % 3 == 0) {
                        throw new IllegalStateException("boom");
                    }
                    return true;
                })
                .subscribe().withSubscriber(AssertSubscriber.create());

        subscriber.request(10)
                .assertFailedWith(IllegalStateException.class, "boom")
                .assertItems(1, 2);
        assertThat(count).hasValue(3);
    }
```

It costs a few extra atomic references for synchronization, but it's more robust than thread sleeps.

Reply:
Maybe it's not the best example :-)

Reply:
This one is better: https://github.com/smallrye/smallrye-mutiny/blob/main/implementation/src/test/java/io/smallrye/mutiny/operators/MultiOnTerminationUniInvokeTest.java#L321